### DISCUSSION TITLE
S3 as document backend

### DISCUSSION THREAD

**Discussion by userbradley** [paperless-ngx/paperless-ngx#762] [Feature Requests]
Howdy,

I've been having a look around and was unable to locate a native solution in Paperless-ngx to use S3 compliant endpoints to store data.

It would be cool if this was part of the container, that if `AWS` credentials are provided, along with a URL, S3 is used.

My work around will probably be something similar to [Using wasabi for Jellyfin](https://breadnet.co.uk/jellyfin-with-s3-backend/) - Where I mount the file system on the host and then pass it's directory to the container.

If you need me to explain this more, please let me know!

**Comment by userbradley** [comment]
Speaking to a colleague about this, I want to add that when I say `s3` I mean `s3 compliant` - so the likes of Minio, AWS S3, Wasabi, etc

**Comment by ryjogo** [comment]
This could be really nice - although it could be achieved easily through external tools, like s3-fuse - https://github.com/s3fs-fuse/s3fs-fuse but with [limitations](https://github.com/s3fs-fuse/s3fs-fuse#limitations). Native support from the application would be great!

**Reply by Crocmagnon** [reply]
I tried s3fuse, and the performance was really poor compared to a local filesystem on the same machine, especially when making FS ops like renaming.
Looking forward to a more native implementation!

**Comment by amo13** [comment]
I guess one could use [django-storages](https://django-storages.readthedocs.io/en/latest/index.html) to implement S3 storage. 

From what I can see, paperless uses [shutil](https://docs.python.org/3/library/shutil.html) for interacting with files. Unfortunately, this does not appear to be the [django default way](https://docs.djangoproject.com/en/4.2/topics/files/#file-storage) to implement file storage. Adding an S3 backend is therefore probably not straightforward. 

I just looked into this because I would like to use minio as an S3 storage backend. But since I have no coding experience with python, I'll have to wait for someone with some python/django experience to give it a try.

**Reply by almereyda** [reply]
There are wrappers around S3 that mimic the shutil API surface, like [s3shutils](https://github.com/andyil/s3shutil), which is reasonably up to date.

**Comment by hegerdes** [comment]
I am also really looking forward for official S3 support. This would make backups much easier (just the DB) and make the deployment of paperless-ngix much more versatile.  You would also get S3 versioning for "free" and it enables so many integration options. 

Personally I'm not a big fan of a S3-fuse since S3 is blob storage not block storage. It is not POSIX compliant and just waiting for a disaster to happen. First class support via  ([boto3](https://github.com/boto/boto3) or [s3shutils](https://github.com/andyil/s3shutil) (uses boto) would be awsome.

**Reply by carvalholeo** [reply]
So, does it just install and change references from shutil to s3shutils and everything will be going to work great? I'm not familiar to Python, so I'm kinda lost here

**Reply by hegerdes** [reply]
That's the aim of s3shutils. But as mentioned before block-storage != blob-storage.  
I think for most pdfs this will not be the problem but what happens if a file is open. How do I represent that in distributed S3. Biggest problem will probably be SQLite. The DB file is always open and written/appended regulary. It will likely be corrupted.

**Reply by userbradley** [reply]
I've got experience running sqlite databases on GCS. My advice is don't. The latency is terrible and the database often gets corrupted

**Reply by carvalholeo** [reply]
Uh, I can see the problem. While I want to use S3 only for documents and use a separate DBMS (as MariaDB or PostgreSQL), I understand that opened documents might be hard to represent into UI os Paperless.

It would be great if Paperless starts to support S3-like by default, only for documents.

Thanks to the answers <3

**Reply by Gsyltc** [reply]
Hi, For a simple usage, S3 can be easily implemented I think.
But the fact that documents are linked to database information can make things much more complex.

Let's imagine the following use case:

S3 (or compatible) can have retention and deletion times. There may even be cases of transfer to another third party (Glacier, etc.).

This implies strong synchronization between the bucket, its data life cycle and the database.

This makes the implementation much more complex than it seems, since a file that moves from one tier to another will have to be deleted from the database. This would require a synchronization mechanism. 
Unless you subscribe to a S3 Message Queue (Like SQS, or other (webhook, amqp, mqtt, etc. for minio for exemple) it would be difficult to know whether it's the bucket that initiates the deletion or tiering.

If you don't do this, you can have an inconsistent database

**Reply by jimkoen** [reply]
To me, the feature request sounds more like something that's understood to be S3 compatible. A typical S3 compatible storage provider doesn't have storage tiers or retention and deletion times. If they can be queried by the regular S3 API, this could still be implemented, but dealing with custom behavior for each S3 provider (S3MQ, webhooks for minio, object lifecycle in Ceph, whatever OpenStack Swift uses) seems very difficult and not worth the effort unless there was a significant part of the user base that uses these features.

For the sake of discussion, is dealing with data lifecycles a hard requirement for people here? If AWS is charging you too much because you're keeping your documents hot, I'd suggest a different S3 compatible storage provider tbh.

**Reply by grapemix** [reply]
> Hi, For a simple usage, S3 can be easily implemented I think. But the fact that documents are linked to database information can make things much more complex.
> 
> Let's imagine the following use case:
> 
> S3 (or compatible) can have retention and deletion times. There may even be cases of transfer to another third party (Glacier, etc.).
> 
> This implies strong synchronization between the bucket, its data life cycle and the database.
> 
> This makes the implementation much more complex than it seems, since a file that moves from one tier to another will have to be deleted from the database. This would require a synchronization mechanism. Unless you subscribe to a S3 Message Queue (Like SQS, or other (webhook, amqp, mqtt, etc. for minio for exemple) it would be difficult to know whether it's the bucket that initiates the deletion or tiering.
> 
> If you don't do this, you can have an inconsistent database

It is very good concern. But if an admin hosts s3 and paperless services. I think the admin should know what he/she doing. We can mention in the doc to disable the data life cycle in the first iteration. We don't have to do everything at once.

The S3 Message Queue is the best solution because it is async, cheap and more bulletproof. But we still need sync solution, scanning whole bucket and rebuilding index/updating db, for SHTF case. Of course, the bucket scanning is expensive if we run it for every update. But I think it is acceptable as a midnight household data integrity checking task. We can just provide a command and let user to fire another docker as a cron job.

For each s3 async notification implementation, I won't suggest we work on it in the first iteration because of the big workload. But we can rely on our community to provide insight or even jump in to help. For example, I am not a ceph expert, but [this page](https://docs.ceph.com/en/quincy/radosgw/notifications/) might be what ceph async notification is about.

**Comment by Gsyltc** [comment]
Hi,
I don't know if I can help but personnaly I use Minio (S3 compatible storage) for my files. 
I use rclone to help me (it's comptabile with Minio, S3, GCP, Azure).

Here the partial compose files to use rclone :
```
version: "3.9"
services:
  webserver:
    image: ghcr.io/paperless-ngx/paperless-ngx:latest
    volumes:
      - type: volume
        source: data
        target: /usr/src/paperless/data
        volume:
          nocopy: true
      - type: volume
        source: media
        target: /usr/src/paperless/media
        volume:
          nocopy: true
      - type: volume
        source: consume
        target: /usr/src/paperless/consume
        volume:
          nocopy: true   
      - export:/usr/src/paperless/export
...

volumes:
  media:    
    driver: rclone
    driver_opts:
      remote: 'homelab:homelab/paperless/media'
      allow_other: 'true'
      vfs_cache_mode: full
      poll_interval: 0
  consume:   
    driver: rclone
    driver_opts:
      remote: 'homelab:homelab/paperless/consume'
      allow_other: 'true'
      vfs_cache_mode: full
      poll_interval: 0 
...

```

You need to have rclone installed and a file for mapping your object storage like
[datas]
type = s3
provider = Minio (or Azure, S3, GCP)
env_auth = false
access_key_id = THE_ACCESS_KEY
secret_access_key = THE_SECRET_KEY
endpoint = THE_ENDPOINT
acl = bucket-owner-full-control
region = THE_REGION

See the documentation of rclone for more information

**Comment by DarkNeedle** [comment]
I’ve seen so many projects like this very away from S3. Implementing S3 is super easy and tying it to the meta data in the database is very easy. Many projects that do not choose to implement S3 or similar storage option and instead defer to fuse or similar shows the inexperience of the developer. PHP is another sign. I may from this and put the S3 option in place. Again s3 has been around a long time and super easy to implement and tie to the meta data.

**Reply by almereyda** [reply]
There's no PHP here, if that would be a blocker for you to contribute voluntarily.

And as a side note, be reassured, that more than half of the sites in the Internet are run on PHP (WordPress) and that there are large installations of PHP which run billions of users (Facebook). The language has also evolved and when used to its full extent, one can build contemporary IT systems with it. I wouldn't have said so in the past myself, but I have learned to give in to admitting the de facto distribution of certain systems.

I'd love to test the branch of a PR that implements S3 support for major providers (AWS, Minio, Ceph).

**Reply by igitur** [reply]
hi @DarkNeedle . I've started work on mounting S3 buckets in the docker container using s3fs. I'm curious to hear about your concerns with using fuse.

**Reply by klafbang** [reply]
@igitur, there are many reasons to not use fuse or similar solutions.  Most important, doing so means the application is not aware it's using S3 storage instead of a file system.  File systems have locking, fast scanning, and incremental updates.  S3 doesn't.

The right way to use S3 is to use it to store data the application never has to refer in detail again.  If you have 1000 documents, a file-system-based application might list the 1000 files, check their modification dates, and read metadata from them.  This is ok on a file system, but disastrous on S3 (slow! and potentially expensive).  Instead, the application should read metadata on upload and store it in a database, then never touch the files again unless they are explicitly requested.  The metadata can be updated for a single file at a time as needed (if the file is later changed), but the original file is never really used by the application, and file collections are never used.

Also, most S3 providers come with built-in CDN, either via public buckets (bad idea for paperless) or via signed URLs.  That means the application does not need to download the file from S3 and then serve it to the browser, but instead provides a single-use time-limited link directly to the browser.  This is faster and reduces load as well as traffic on the application server.

Using S3 via fuse cannot have these advantages and will inevitably break once you have a couple thousand files.

**Comment by tb1337** [comment]
Looking forward this gets implemented.

**Reply by Atomique** [reply]
Same here.

**Comment by grapemix** [comment]
I would like to provide additional benefits for switching to object storage in case someone asked in the future.

    Build-in object tagging
    Build-in quota support
    Build-in event support
    Build-in versioning in object level
    Third-party integrated backup solution
    Build-in ACL
    Has third-party IAM liked permission system (but it is complicated to setup)
    Redundancy (not just in disk level, but also instance level)
    Easy horizontal scalability (not just in disk level, but also instance level)
    Easy size expansion (we don't have to copy files if we want to expand the storage)
    Able to share file in limited time (like via pre-signed url)
    Easy to find library to upload from client side (like via pre-signed url)
    Designed for parallel IO. Ceph split files into multiple drives, so workloads will also separated into multiple drives instead of one.

If we have truly object storage layer, the docker container can become truly stateless which means we can have multiple k8s pods and spin up and down if we need to among different instances. It is painful to share persistent volume between different pods/projects.... Yes, we can hack it, but it's painful to watch ;)

Is it overkill? probably, but we can also simply use midnight commander for our doc management too if we are minimalist , right? ;)

Supporting object storage sounds like an investment to me. Yes, we have to spend some resources on this feature, but this features can save us lots of time in the futures because of the benefits being shown above.

**Reply by jimkoen** [reply]
Not only is it overkill, it would likely break backwards compatibility. And paperless already has rudimentary support for RBAC, when talking in terms of access control.

**Reply by grapemix** [reply]
1. We don't have to use s3's RBAC if it is the blocker. RBAC is just one of many features s3 layer offers.
2. Can you elaborate what backwards compatibility do you mean? Here is a few generic solutions for the common backwards compatibility:
   a. Add s3 storage instead of replacing fs storage. Admin can choose via cfg/env var.
   b. Add new db col for s3 data instead of using old db col
   c. For data migration:
         c1. Either only allow using s3 storage for new install
         c2. Or writing script to migrate files from fs to s3 and update the db table. May be in next stage, so users have to wait.
         c3. Each solution has trade-off. Instead of guess what users want, why do you let users choose? May be someone will jump in to help.
 3. We can always divide and conquer a big problem. Instead of worrying we don't have fully S3 compatibility, we can start with minimum set of S3 compatibility first.

**Reply by jimkoen** [reply]
> Can you elaborate what backwards compatibility do you mean?

I was referring to your suggestion of making use of S3's specific ACL implementation and semantics. Without knowing the specifics of how S3 does ACL, zeroing in on a specific implementation would break or change compatibility for users that want to remain on a regular FS storage backend.

I also believe that the majority of the community is not requiring S3 compatible storage for their personal document management system, so an implementation should be careful about introducing new features specificially for S3 based storage backends. S3 storage seems like a hardcore enthusiast / commercial use case anyways, forcing users to switch would make no sense.

**Comment by ajfriesen** [comment]
You can use sqlite and then use the project https://litestream.io/ to stream your backup to any s3 compatible storage.

**Comment by CoE2013** [comment]
Hi Guys,

we have started using paperless and use S3 Storage as descriped by mounting via rclone.
DB is running locally (due to performance in the future), but we back it up to s3 as well... 

For us, seems like a good working solution, except this issue:

- upload a doc
- assign another storage path
=> this should be refelected, but instead, the org file gets deleted !

after digging around a bit, it seems that S3 storage in gerneal does not support moving files, but instead we need a cp / delete operation.
Can anyone confirm?

Also, does paperless mv files to sync changed storage paths?
Could this be made a cp and delete instead - this should probably fix this issue and then we have a nice working solution using S3 (even with encryption, versionening and all the stuff)

**Reply by CoE2013** [reply]
Hi all,

can anyone point us to the right file / line where the storage path is set / updated?
In handlers.py we found

```
document.storage_path = selected
document.save(update_fields=("storage_path",))
```

but where is the file on the HDD actually moved?

**Comment by almereyda** [comment]
Would you like to open a separate discussion for this issue of an
officially unsupported configuration?

**Comment by CoE2013** [comment]
sure
[[](url)](https://github.com/paperless-ngx/paperless-ngx/discussions/8781)

**Comment by Daemonslayer2048** [comment]
Like others in this thread I also would like a native S3 implementation. My implementation differs from what I have seen discussed above though, rclone to s3 does not work in my use case as I deploy to a Kubernetes cluster. Its possible I could do something exceptionally hack-y like using a hostpath to an rclone directory on the host but thats not a fun idea. Just tossing in my two cents on what would be a nice improvement to the project in my opinion.

**Reply by userbradley** [reply]
@Daemonslayer2048 you can build the image with s3 fuse in and use an init command. Did something similar for GCS a while ago. Happy to share notes as a slightly less hacky solution, but still hacky

**Reply by Atomique** [reply]
@userbradley do you mind me to ask for your implementation? Would be interested into it.

**Reply by userbradley** [reply]
Yeah so I've done this for GCS (google cloud storage) for other apps but it would work the same I _assume_ for S3

Bake https://github.com/s3fs-fuse/s3fs-fuse in to your docker image, so it would just be using the paperless-ngx image as the from, then copy it to `/bin` or where ever.

Then for the k8s stuff, something like this would work where you call a post start hook which mounts the s3 bucket as a file system

https://github.com/userbradley/gcs-web-server/blob/main/helmcharts/gcs-web-server/templates/deployment.yml

An example of doing something similar with GCS and nginx is here: https://github.com/userbradley/gcs-web-server/blob/main/container/Dockerfile

**Reply by Atomique** [reply]
Dont you think it would be possible to add a s3fuse container as second Sidecar container into the pod and mount it into the pod as volume?

**Reply by userbradley** [reply]
Ive not tried using a sidecar container for this, but _in theory_ it should work as they share volumes, but my thoughts around not using it is the file permissions between containers. I'd say give it a go, but I remember having a think about this and not doing it for _some reason_

**Comment by iedame** [comment]
I also would be interested in this feature.

**Comment by Maddin-619** [comment]
I would also love to have native remote filesystem support. But in my case it is FTP.
What do you guys think about implenting an fs abstraction layer for example by using [fsspec](https://github.com/fsspec/filesystem_spec)? This gives us not only S3 but many other file systems as well.

**Comment by farezascodes** [comment]
Hello @shamoon ,
I’d like to contribute to this feature request. I have implemented S3-compatible storage using python django-storage lib for Paperless (tested with MinIO). This implementation has been running in my company’s production environment for over a year. Currently it holds about 3.4K files that are accessed daily.

The core functionality that relies on the filesystem remains unchanged, the implementation only switches the final storage location from local disk to S3 for persisted documents. The flow I implemented is as follows: <br>
<img width="512" height="384" alt="image" src="https://github.com/user-attachments/assets/522a5ffd-c683-4c33-af7c-13e57229915d" />

<br>
I’d greatly appreciate any feedback or suggestions regarding this approach.

**Reply by shamoon** [reply]
Your code doesn't seem available online so not much feedback to provide. In general my only hesitation with something like this is the size/scope and thus, risk and of course the support burden.

This being open source, users are always welcome to contribute (understanding that merging isn't "guaranteed").

**Reply by farezascodes** [reply]
Yes currently the code is in my company repository and uses older version of paperless. Maybe i will try to integrate it to the newest version first so that people can access it in public. 
I will inform you when it is available. Thank youu

**Reply by klafbang** [reply]
Re: your flow:

1) consider whether it is necessary to make the upload synchronously.  Why not do OCR + upload to S3 in the background so the user doesn't have to wait?
2) consider using pre-signed URLs for uploads instead: send a request to the backend, which creates an entry in the DB and makes a pre-signed upload URL, the frontend then uploads directly to S3.  After successful upload, the frontend can notify the backend, or the backend can periodically check for itself, and when the upload is complete it can do OCR and update the status.  This saves round-trips frontend -> backend -> s3, reducing the backend load
3) you don't show how you handle the display of documents in S3.  The obvious, and sub-optimal, way is to fetch from the backend, which fetches from S3.  A much better solution is to send a (pre-signed) S3 URL directly to the frontend; that way, the backend never sees uploads again, reducing traffic and load

MinIO and most cloud-providers support pre-signed URLs: http://mindev.org/presigned-put-upload-via-browser.html

**Reply by farezascodes** [reply]
Thank you for the feedback. Yes i agree with your approach. But maybe i will replicate my approach first and then we can discuss it together for the improvement. But here is my concern:
1. I do it synchronously because i don't want to break the core functionality / business flow too much, because of development time XD. And also i don't want to handle the needs to delete the file if OCR is failed, which is paperless emphasize (the OCR function) CMIIW
2. Got it, maybe i will consider it after the initial feature release!!
3. Yes agree, my implementation use the round trip between frontend -> backend -> s3 and it is bandwidth heavy XD. But at that time, the bandwidth is not the issue because i am using on premise server with zero cost for the network. But i will make the pre-signed URL as the top priority feature then

ps. this is my image getter flow XD
<img width="584" height="384" alt="image" src="https://github.com/user-attachments/assets/c63f3241-dfa8-4212-aca4-be8771c9cea2" />

**Reply by BenRoe** [reply]
@farezascodes  Does you code support "object lock" with Retention (Governance Mode)? Would be great for gobd konform.
https://docs.hetzner.com/storage/object-storage/faq/buckets-objects#what-is-the-difference-between-versioning-and-object-locking

That would be a massive feature for german user, because of GoBD-konform document storage.

**Reply by farezascodes** [reply]
@BenRoe my codes not yet supports that. I only tried to change the storage location for now, not implementing the full features of s3 XD

**Reply by kemboi22** [reply]
Hello @farezascodes Hope You are Good
Have you been able to impliment the S3 implimentation

**Comment by Pascal-Thiel** [comment]
@kemboi22 Since April there is a new AWS S3 Feature: S3Files. With this you can mount s3 as a file System.
It does not store the data directly on S3. There is a files system for local data access. S3 will synchronize each File after 60 Seconds file-inactivity.
We simply mounted the Paperless folders on S3Files. Works without any issues. No modification in paperless needed.
https://aws.amazon.com/de/s3/features/files/

**Reply by hegerdes** [reply]
While being 100% right, this is kind of a short-sighted solution.  
S3Files is a AWS only service that builds on top of S3 and EFS, it limits users to AWS and can cause additional cost depending on the applications R/W pattern.  

S3Files should be treated as a AWS specific workaround, not a solution for a native implementation as requested here.

**Reply by kemboi22** [reply]
But this limits one on AWS and for me i would like if one can use it on opensource S3 servers like RUSTFS
@Pascal-Thiel  thanks for the suggestion will check it out and explore it

**Reply by stex** [reply]
In the end, the `rclone` docker plugin solution (https://github.com/paperless-ngx/paperless-ngx/discussions/762#discussioncomment-8018419) is pretty much what AWS offers now.
